### ISSUE TITLE
jax.experimental.rnn has NaNs in gradients

### ISSUE THREAD

**Issue by sourabh2k15** [google/jax#15183]
### Description

JAX experimental CuDNN backed LSTM added in https://github.com/google/jax/pull/13319 
has the following issues : 

1) NaNs in backward pass. 
2) if bidirectional = True, forward pass values between [lstm_ref](https://github.com/google/jax/blob/main/jax/experimental/rnn.py#L254) and [lstm](https://github.com/google/jax/blob/main/jax/experimental/rnn.py#L221) do not match. This is after ignoring the padded parts. 

Reproduction : 

Run this code on any google colab with GPU hosted runtime that uses a T4 GPU 
or VM with CUDA 11, CuDNN 8.5, jax 0.4.1, jaxlib 0.4.1+cuda11.cudnn8.2 GPU V100

```
import numpy as np
import jax
import jax.numpy as jnp
from jax.experimental import rnn


batch_size = 8
input_size = 8
hidden_size = 4
num_layers = 1
bidirectional = True
num_directions = 2 if bidirectional else 1

seq_lengths = jnp.array([4, 5, 4, 1, 1, 1, 1, 1], dtype=jnp.int32)
max_seq_length = 5

print(seq_lengths)
print(jax.local_device_count())

root_key = jax.random.PRNGKey(1)
k1, k2, k3, k4 = jax.random.split(root_key, 4)
x = jax.random.normal(
    k1, (batch_size, max_seq_length, input_size), dtype=jnp.float32)
h_0 = jax.random.normal(
    k2, (num_directions * num_layers, batch_size, hidden_size),
    dtype=jnp.float32)
c_0 = jax.random.normal(
    k3, (num_directions * num_layers, batch_size, hidden_size),
    dtype=jnp.float32)
weights = rnn.init_lstm_weight(k4, input_size, hidden_size, num_layers,
                                bidirectional)

def f(x, h_0, c_0, weights):
  y, h, c = rnn.lstm(
      x,
      h_0,
      c_0,
      weights,
      seq_lengths=seq_lengths,
      input_size=input_size,
      hidden_size=hidden_size,
      num_layers=num_layers,
      dropout=False,
      bidirectional=bidirectional)
  loss = jnp.sum(y)
  return loss, (y, h, c)

# y, h_n, c_n = f(x, h_0, c_0, weights)

grad_fn = jax.value_and_grad(f, has_aux=True)
(loss, (y, h_n, c_n)), grad = grad_fn(x, h_0, c_0, weights)
```

I get the following output : 
```
[4 5 4 1 1 1 1 1]
1
Are NaNs present =  True
Number of NaNs =  320
```

[Google Colab](https://colab.sandbox.google.com/drive/1O6EEK_0OMCv5vfEoWruDzPg1xuvDBpCj?resourcekey=0-QyZUpUXDOEFKpDGgwTtsyQ#scrollTo=6x0Qsci8Sega)

### What jax/jaxlib version are you using?

jax 0.4.1 , jaxlib 0.4.1+cuda11.cudnn82

### Which accelerator(s) are you using?

GPU

### Additional system info

Ubuntu 20.04 , V100 GPU 

### NVIDIA GPU info

Run on google colab 

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   33C    P8     9W /  70W |      0MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by hawkinsp** [comment]
I think #15213 fixed this.
